torch/_dynamo/utils.py

```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
         # ignore this case of getattr
         getattr_fn = None
     return getattr_fn
+
+
+# Note - we anticipate changes here in the future, and so the reasoning of
+# the util feels worth not replicating this logic in multiple places,
+# even if the return type is really ugly.
+def tensor_shape_should_be_static(
+    tensor: Union[torch.Tensor, Any], source: Optional["Source"], is_tensor: bool
+) -> Tuple[bool, str]:
+    """
+    Given a tensor, source, and is_tensor flag, determine if a shape should be static.
+
+    Args:
+    tensor - the tensor to evaluate, parameters force a static shape.
+    source - an optional source, None forces a static shape
```

Comment:
Do you remember when `wrap_to_fake_tensor_and_record` can be given a None source?

Reply:
No, let me check

Reply:
stacked a PR on top, I didn't see anything in tests, but there may be some part of CI that triggers it.

Reply:
may i recommend good old fashioned code reading